vpr/src/place/initial_placement.cpp

```diff
@@ -122,6 +215,84 @@ static bool is_block_placed(ClusterBlockId blk_id) {
     return (!(place_ctx.block_locs[blk_id].loc.x == INVALID_X));
 }
 
+static int get_y_loc_based_on_macro_direction(t_grid_empty_locs_block_type first_macro_loc, t_pl_macro pl_macro) {
+    int y = first_macro_loc.first_avail_loc.y;
+
+    /*
+     * if the macro member offset is positive, it means that macro head should be placed at the first location of first_macro_loc.
+     * otherwise, macro head should be placed at the last available location to ensure macro_can_be_placed can check macro location correctly.
+     * 
+     */
+    if (pl_macro.members.size() > 1) {
+        if (pl_macro.members.at(1).offset.y < 0) {
+            y += (pl_macro.members.size() - 1) * abs(pl_macro.members.at(1).offset.y);
+        }
+    }
+
+    return y;
+}
+
+static void get_blk_type_first_loc(t_pl_loc& loc, t_logical_block_type_ptr block_type, t_pl_macro pl_macro) {
+    const auto& device_ctx = g_vpr_ctx.device();
+
+    //loop over all empty locations and choose first column that can accomodate macro blocks
+    for (unsigned int empty_loc_index = 0; empty_loc_index < blk_types_empty_locs_in_grid[block_type->index].size(); empty_loc_index++) {
+        auto first_empty_loc = blk_types_empty_locs_in_grid[block_type->index].at(empty_loc_index);
+
+        //if macro size is larger than available locations in the specific column, should go to next available column
+        if ((unsigned)first_empty_loc.num_of_empty_locs_in_y_axis < pl_macro.members.size()) {
+            continue;
+        }
+
+        //set the coordinate of first location that can accomodate macro blocks
+        loc.x = first_empty_loc.first_avail_loc.x;
+        loc.y = get_y_loc_based_on_macro_direction(first_empty_loc, pl_macro);
+        loc.sub_tile = first_empty_loc.first_avail_loc.sub_tile;
+
+        //update the first available macro location in vector for the next macro
+        first_empty_loc.first_avail_loc.y += device_ctx.physical_tile_types.at(block_type->index).height * pl_macro.members.size();
```

Comment:
Seems like the update routine should be called only if the placement of this macro succeeds (and there are floorplan etc. checks to do still).  So I think you should separate this into two routines: get_next_free_location and update_free_location_list (or some names like that).